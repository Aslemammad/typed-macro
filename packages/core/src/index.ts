export * from './env'
export * from './macro'
export * from './exports'
export * from './provider'
export * from '../types'
