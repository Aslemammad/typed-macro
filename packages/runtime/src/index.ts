export * from './transformer'
export * from './normalizer'
export * from './renderer'
export * from './runtime'
export * from './loader'
export * from './filter'
export * from './hook'
