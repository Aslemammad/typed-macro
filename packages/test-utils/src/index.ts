export * from './transformer'
export * from './runtime'
export * from './renderer'
export * from './env'
