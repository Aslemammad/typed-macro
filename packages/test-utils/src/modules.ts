import { Modules } from '@typed-macro/core'
import { isString } from '@typed-macro/shared'

/** @internal */
export function createMockModules(): Modules {
  const container: Map<string, string> = new Map()

  function query(pattern: RegExp | string) {
    const result: string[] = []
    const checker = isString(pattern)
      ? (tag: string, id: string) => pattern === tag && result.push(id)
      : (tag: string, id: string) => pattern.test(tag) && result.push(id)
    container.forEach(checker)
    return result
  }

  return {
    getTag(id) {
      return container.get(id)
    },
    setTag(id, tag) {
      container.set(id, tag)
    },
    unsetTag(id) {
      container.delete(id)
    },
    queryByTag: query,
    invalidateByTag: query,
  }
}
