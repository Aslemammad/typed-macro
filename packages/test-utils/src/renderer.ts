import { ModularizedExportable } from '@typed-macro/core'
import { createTypeRenderer, normalizeExports } from '@typed-macro/runtime'

export type TestTypeRenderer = (exports: ModularizedExportable) => string

/**
 * Create a type renderer to help test the type of macros/modules.
 *
 * e.g.
 * ```typescript
 * const aMacro = defineMacro(...)
 *   .withSignature(...)
 *   .withHandler(...)
 *
 * const renderType = createTestTypeRenderer()
 *
 * expect(
 *   renderType({
 *     '@ns': {
 *       macros: [aMacro],
 *       types: `some types...`,
 *     },
 *     '@another': {
 *       code: `some code...`,
 *       types: `some types...`,
 *     }
 *   })
 * ).toMatchSnapshot()
 * ```
 */
export function createTestTypeRenderer(): TestTypeRenderer {
  return (exports) => {
    const renderer = createTypeRenderer()
    const { types } = normalizeExports(exports)
    Object.keys(types).forEach((moduleName) =>
      renderer.append(moduleName, types[moduleName])
    )
    return renderer.renderToString()
  }
}
