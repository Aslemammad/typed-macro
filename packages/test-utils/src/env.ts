import { EnvContext, FSWatcher, Modules, WatchOptions } from '@typed-macro/core'
import chokidar from 'chokidar'
import { createMockModules } from './modules'

export type TestEnvContextOptions = Partial<
  Omit<EnvContext, 'watcher' | 'modules'>
> & {
  /**
   * Options for EnvContext.watcher.
   *
   * If is undefined, then EnvContext.watcher will be undefined.
   */
  watchOptions?: WatchOptions

  /**
   * Options for EnvContext.modules.
   *
   * If is true, then EnvContext.modules will be a mocked Modules.
   * If is undefined, then EnvContext.modules will be undefined.
   */
  modules?: boolean | Modules
}

/**
 * Create EnvContext for testing.
 *
 * Options will be merged with the default env.
 *
 * Default env:
 * ```typescript
 * {
 *    projectPath: ['/'],
 *    packageManager: 'test',
 *    host: 'test',
 *    dev: true,
 *    ssr: false,
 *    modules: undefined,
 *    watcher: undefined
 * }
 * ```
 */
export function createTestEnvContext({
  modules,
  watchOptions,
  ...env
}: TestEnvContextOptions = {}): EnvContext {
  return {
    projectPath: ['/'],
    packageManager: 'test',
    host: 'test',
    dev: true,
    ssr: false,
    modules: modules
      ? modules === true
        ? createMockModules()
        : /* istanbul ignore next */ modules
      : undefined,
    watcher:
      watchOptions && (new chokidar.FSWatcher(watchOptions) as FSWatcher),
    ...env,
  }
}
