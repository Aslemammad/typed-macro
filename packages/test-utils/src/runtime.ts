import { EnvContext, MacroProvider } from '@typed-macro/core'
import { createRuntime, Runtime, RuntimeOptions } from '@typed-macro/runtime'
import { isArray } from '@typed-macro/shared'
import { createTestEnvContext } from './env'

export type TestRuntimeOptions = RuntimeOptions & {
  /**
   * EnvContext for testing.
   *
   * If is undefined, use the default one created by {@link createTestEnvContext}.
   */
  env?: EnvContext
  /**
   * Providers to be appended into runtime.
   */
  provider: MacroProvider | MacroProvider[]
}

export type TestRuntime = Omit<Runtime, 'renderTypes' | 'appendProvider'> & {
  /**
   * Render types into string.
   */
  renderTypes(): string
}

/**
 * Create a runtime to help test providers.
 *
 * e.g.
 * ```typescript
 * const provider = defineMacroProvider(...)
 *
 * const runtime = createTestRuntime({
 *   provider
 * })
 *
 * expect(
 *   runtime.renderTypes()
 * ).toMatchSnapshot()
 *
 * expect(
 *   runtime.load('ns')
 * ).toMatchSnapshot()
 *
 * expect(
 *   runtime.transform(`code`, 'filename')
 * ).toMatchSnapshot()
 *
 * // and so on...
 * ```
 */
export function createTestRuntime({
  provider,
  env = createTestEnvContext(),
  ...runtimeOptions
}: TestRuntimeOptions): TestRuntime {
  const runtime = createRuntime(env, runtimeOptions)

  if (isArray(provider)) provider.forEach((p) => runtime.appendProvider(p))
  else runtime.appendProvider(provider)

  return {
    ...runtime,
    renderTypes: () => runtime.internal.renderer.renderToString(),
  }
}
