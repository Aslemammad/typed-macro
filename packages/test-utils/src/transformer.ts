import { EnvContext, Macro } from '@typed-macro/core'
import { createTransformer, TransformerOptions } from '@typed-macro/runtime'
import { isString } from '@typed-macro/shared'
import { createTestEnvContext } from './env'

export type TestTransformerContext = {
  /**
   * Source code.
   */
  code: string

  /**
   * Path of the source file.
   *
   * Default to 'test.ts'
   */
  filepath?: string

  /**
   * EnvContext for testing.
   *
   * If is undefined, use the default one created by {@link createTestEnvContext}.
   */
  env?: EnvContext
}

export type TestTransformer = {
  (ctx: TestTransformerContext): string
  (code: string): string
}

export type TestTransformerOptions = TransformerOptions & {
  /**
   * Macros to be appended into transformer.
   */
  macros: { [moduleName: string]: Macro[] } // Keep ModularizedMacros internal.
}

/**
 * Create a transformer to help test macros.
 *
 * e.g.
 * ```typescript
 * const aMacro = defineMacro(...)
 *   .withSignature(...)
 *   .withHandler(...)
 *
 * const transform = createTestTransformer({
 *   macros: { '@ns': [aMacro] },
 * })
 *
 * expect(
 *   transform(`some code...`)
 * ).toBe(`some code...`)
 * ```
 */
export function createTestTransformer({
  maxTraversals,
  parserPlugins,
  macros,
}: TestTransformerOptions): TestTransformer {
  const transformer = createTransformer({ maxTraversals, parserPlugins })
  Object.keys(macros).forEach((moduleName) =>
    transformer.appendMacros(moduleName, macros[moduleName])
  )
  return (input) => {
    // normalized
    const {
      code,
      filepath = 'test.ts',
      env = createTestEnvContext(),
    } = isString(input) ? { code: input } : input

    // prepare transformer
    return transformer.transform(code, filepath, env) || code
  }
}
