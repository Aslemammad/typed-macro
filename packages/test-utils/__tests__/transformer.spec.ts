import { createTestEnvContext, createTestTransformer } from '../src'
import { createMacroRemoveOnProd } from '../../core/__tests__/macros'

describe('TestTransformer', () => {
  it('should work', () => {
    const transform = createTestTransformer({
      macros: {
        '@macros': [createMacroRemoveOnProd()],
      },
    })

    const code = `
    import { dev } from '@macros'
    let a = 0
    dev(a)
    `

    const inProd = transform({
      code,
      env: createTestEnvContext({ dev: false }),
    })
    const inDev = transform(code)

    expect(inProd).not.toBe(inDev)
    expect(inProd).toMatchSnapshot()
    expect(inDev).toMatchSnapshot()
  })
})
