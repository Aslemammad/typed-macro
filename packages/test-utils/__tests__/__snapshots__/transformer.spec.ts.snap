// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestTransformer should work 1`] = `
"

let a = 0;"
`;

exports[`TestTransformer should work 2`] = `
"

let a = 0;
a;"
`;
